### PR TITLE
Restore collapsers

### LINKED ISSUE #679: CM6: Restore collapser
This is a follow-up of https://github.com/jupyter/nbdime/pull/673 in which the collapser of unchanged code was removed (as not ported from CM5 to CM6).

Some solution could probably be extrapolated from https://github.com/codemirror/merge

### PR DESCRIPTION
**Pull request by fcollonval**
Fixes #679

### CODE REVIEW

**Comment by fcollonval** [comment]
bot please update snapshots

**Comment by fcollonval** [comment]
Kicking the CI

File: packages/nbdime/src/common/mergeview.ts
Comment (HaudinFlorence) [inline packages/nbdime/src/common/mergeview.ts:1]: In mergeview.ts, in `collapseIdenticalStretches`, maybe `edit` could be replaced by a more specific name like:
```const edit = this.base.cm;```=> ```const baseEditor = this.base.cm;```

File: packages/nbdime/src/common/mergeview.ts [hunk 1]
```diff
@@ -409,6 +421,92 @@ const conflictMarkerLineChunkMappingField = StateField.define<
...
+ * Uncollapse effect to synchronize uncollapsed ranges between editors.
+ */
+const uncollapse = StateEffect.define<number>({
+  map: (value, change) => change.mapPos(value),
+});
+
+/**
+ * Effect to set the collapser widgets in an editor
+ */
+const setCollapsers = StateEffect.define<DecorationSet>({
```
Comment (HaudinFlorence) [inline packages/nbdime/src/common/mergeview.ts:434]:
```suggestion
 */
const setCollapsersEffect = StateEffect.define<DecorationSet>({
```
All effects have `Effect` as a suffix.

File: packages/nbdime/src/common/mergeview.ts [hunk 2]
```diff
@@ -409,6 +421,92 @@ const conflictMarkerLineChunkMappingField = StateField.define<
   },
 });
 
+/**
+ * Uncollapse effect to synchronize uncollapsed ranges between editors.
+ */
+const uncollapse = StateEffect.define<number>({
```
Comment (HaudinFlorence) [inline packages/nbdime/src/common/mergeview.ts:427]:
```suggestion
const uncollapseEffect = StateEffect.define<number>({
```

File: packages/nbdime/src/common/mergeview.ts [hunk 3]
```diff
@@ -1471,6 +1612,92 @@ export class MergeView extends Panel {
...
   }
 
+  protected collapseIdenticalStretches(): void {
+    const margin = this._collapseIdentical;
+    if (margin < 0) {
+      return;
+    }
+
+    // Build an array of line that are not part of a chunks
+    const edit = this.base.cm;
```
Comment (HaudinFlorence) [inline packages/nbdime/src/common/mergeview.ts:1622]:
```suggestion
    const baseEditor = this.base.cm;
```

File: packages/nbdime/src/common/mergeview.ts [hunk 4]
```diff
@@ -123,14 +123,26 @@ const mergeViewDecorationDict: MergeViewDecorationDict = {
...
+ */
+const baseTheme = EditorView.baseTheme({
+  '.cm-collapsedLines': {
+    padding: '5px 5px 5px 10px',
+    cursor: 'pointer',
+    color: 'var(--jp-ui-font-color2)',
+    outline: 'solid var(--jp-border-width) var(--jp-border-color1)',
+  },
+});
+
```
Comment (HaudinFlorence) [inline packages/nbdime/src/common/mergeview.ts:137]:
```suggestion
const baseTheme = EditorView.baseTheme({
    cursor: 'pointer',
    outline: 'solid var(--jp-border-width) var(--jp-border-color1)',
    color: 'var(--jp-ui-font-color1)',
    backgroundColor: 'var(--jp-layout-color2)',
    border: '1px solid var(--jp-border-color1)',
    fontSize: '90%',
    padding:'0 3px',
    borderRadius: '4px'
  },
});

```
The suggestion proposed here is the following one or the rendering of the collapsers:
Screenshot from 2023-10-03 16-51-58

File: ui-tests/tests/nbdime-merge-test10.spec.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+test.beforeEach(async ({ page }) => {
+  await page.goto('http://localhost:41000/merge');
+  await page.locator('#merge-local').fill('data/merge_test10/left.ipynb');
+  await page.locator('#merge-base').fill('data/merge_test10/center.ipynb');
+  await page.locator('#merge-remote').fill('data/merge_test10/right.ipynb');
+  await page.getByRole('button', { name: 'Merge files' }).click();
+});
+
+/* notebooks of same length and 1 conflict*/
+test.describe('merge test10', () => {
```
Comment (HaudinFlorence) [inline ui-tests/tests/nbdime-merge-test10.spec.ts:12]:
```suggestion
test.describe('merge test4', () => {
```

Also the indexation of examples could start from 1.

File: packages/nbdime/src/common/mergeview.ts
Comment (fcollonval) [reply packages/nbdime/src/common/mergeview.ts:1]: thanks for the review Florence

**Comment by fcollonval** [comment]
bot please update playwright snapshots

**Review by HaudinFlorence — APPROVED** [review]

File: ui-tests/tests/nbdime-diff-test3.spec.ts-snapshots/diff-test3-take-a-snapshot-at-opening-1-win32.png
Comment (vidartf) [inline ui-tests/tests/nbdime-diff-test3.spec.ts-snapshots/diff-test3-take-a-snapshot-at-opening-1-win32.png:1]: The windows version retains the lock spacer, while the linux version does not?

File: ui-tests/tests/nbdime-merge-test4.spec.ts-snapshots/merge-test4-should-synchronize-the-collapse-status-between-editor-1-win32.png
Comment (vidartf) [inline ui-tests/tests/nbdime-merge-test4.spec.ts-snapshots/merge-test4-should-synchronize-the-collapse-status-between-editor-1-win32.png:1]: The linux version has `... X unchanged lines ...` while Windows version does not have ellipsis?

**Review by vidartf — COMMENTED** [review]
Haven't looked to closely at this yet, but a skim through showed some weird differences between linux and Windows UI test screenshots?

File: ui-tests/tests/nbdime-merge-test4.spec.ts-snapshots/merge-test4-should-synchronize-the-collapse-status-between-editor-1-win32.png
Comment (HaudinFlorence) [reply ui-tests/tests/nbdime-merge-test4.spec.ts-snapshots/merge-test4-should-synchronize-the-collapse-status-between-editor-1-win32.png:1]: @vidartf Indeed that's strange.

**Comment by krassowski** [comment]
The issue with divergent snapshots is because one of the updates failed (the Linux one) because between starting the job and finishing it someone pushed another commit to the branch which prevented it from pushing the udpated screenshots https://github.com/jupyter/nbdime/actions/runs/6420290475/job/17432051180. Because the jobs run sequentially the other one did not fail because it already fetched the updated git repo. To prevent inconsistent state we could make the entire job fail if any of them fails. For now:


bot please update playwright snapshots

**Comment by HaudinFlorence** [comment]
bot please update playwright snapshots

File: packages/nbdime/src/common/mergeview.ts [hunk 6]
```diff
@@ -409,6 +424,92 @@ const conflictMarkerLineChunkMappingField = StateField.define<
...
+              otherSibling.cm == sibling.cm,
+          ) >= 0,
+      )
+    );
+  }
+
+  toDOM(view: EditorView) {
+    let outer = document.createElement('div');
+    outer.className = 'cm-collapsedLines';
+    outer.textContent = view.state.phrase('...$ unchanged lines...', this.lines);
```
Comment (vidartf) [inline packages/nbdime/src/common/mergeview.ts:466]: If we want this to look similar to older versions of nbdime, we could simply have this be `(...)`. Maybe having `(...) $ unchanged lines` ? Mostly subjective taste at this point, so I'm happy to leave the final judgment to you.

Comment (fcollonval) [inline packages/nbdime/src/common/mergeview.ts:466]: @ vidartf I'm changing for the following:

```suggestion
    outer.textContent = view.state.phrase('(…$ unchanged lines…)', this.lines);
```

**Comment by fcollonval** [comment]
bot please update playwright snapshots

**Review by krassowski — APPROVED** [review]
LGTM, thanks!

**Pull request merged**